### ISSUE TITLE
Make cache thread-local

### ISSUE THREAD

**Issue by mcpl-sympy** [sympy/sympy#18210]
Please refer to PR #18198.

SymPy's global parameters are now thread-local, but cache is not. This leads to undesired result in multi-threading process.

I have tried to make cache thread-local by modifying `sympy.core.cache` like this:
``` python
>>> import threading
>>> ns = threading.local()
>>> ns.CACHE = _cache()
```
However, this was not working. Cache in one thread was still intefering the other.
Any suggestion on this, please?

**Comment by oscarbenjamin** [comment]
It's hard to say from your description exactly what you've done but to make the cache thread-safe we need to make it so that every time an object is looked up in the cache the thread-local cache is looked up first. That requires some rearrangement in the cache code.

**Comment by mcpl-sympy** [comment]
@oscarbenjamin 

Seeing now, my previous description seems wrong. 
I have a vague blueprint in mind, but don't know how to implement it exactly.

Just as `sympy.core.parameters._global_parameters` is subclass of `threading.local`, `sympy.core.cache._cache` must be subclass of `threading.local` as well, I reckon.

Now, since instance of `threading.local` stores the thread-safe variables as attribute, `sympy.core.cache.CACHE` must store the cached functions with dict-like behavior.

However, current `sympy.core.cache.CACHE` stores them with list-like behavior, and I don't know how to make this transition seamlessly. This whole module is too abstruse for me.

**Comment by oscarbenjamin** [comment]
Maybe for each cached function a `threading.local` should be created that stores the `lru_cache` of that function as a thread-local attribute. Then the cache ends up being a list of `threading.local` objects. The `wrapper` function will need to retrieve (or create) the thread-local `lru_cache` when called.

**Comment by mcpl-sympy** [comment]
@oscarbenjamin 
I have tried several more approaches, but all have failed.  
It seems that `fastcache.clru_cache` refers to some internal container. (fastcache module is implemented in C.)  
Removing the dependency on fastcache might do, but it will slow down the overall performance.

**Comment by oscarbenjamin** [comment]
Since `fastcache` is an optional dependency can we make the fallback cache threadsafe?

**Comment by mcpl-sympy** [comment]
> Since fastcache is an optional dependency can we make the fallback cache threadsafe?

I believe so, since for other caches we can get direct access to every container where the cached functions, arguments, and results are stored.

**Comment by mcpl-sympy** [comment]
For anyone who might delve into this in the future, I'll write down what have been tried and why they have failed.  
* Disclaimer: I hold no expertise in cacheing and threading. The contents below needs to be reviewed by more skillful contributer.

# Summary

## What's the matter here?
- In PR #18198, SymPy's global parameters are made thread-local.  
This means that if we set `global_parameters.evaluate=True` for thread A and `global_parameters.evaluate=False` for thread B and run `x+x` parallelly, the former should return `2*x` while the latter returns `x+x`.
- But actually, it doesn't. That's because `Add(x, x)` is cached. Once `2*x` is returned from thread A, this result is stored and returned when `x+x` is run in thread B.

## How can we achieve thread-safety of the cache?
We need to maintain different caches for different threads. Thread A must refer to cache A only, and for thread B, cache B.

## Why have you failed?
- SymPy's cacheing systems chooses one of these three caching function: `fastcache.clru_cache`, `functools.lru_cache`, `sympy.core.compatibility.lru_cache`, in the order of decresing preference.  
- Since the first two are outside of SymPy, we can't manipulate their containers. Even if we make the list of cacheing functions thread-local, the containers they store their results and refer to remain thread-global.
- We can easily scrap `functools.lru_cache` and use `sympy.core.compatibility.lru_cache`, which can be modified at our will. The problem is `fastcache.clru_cache`. This cannot be naively supplanted, since it runs much faster than other modules because it is C-implementation.

**Comment by oscarbenjamin** [comment]
I don't think we need to scrap `functools.lru_cache`. We can just make a separate cache object for each function for each thread. When the wrapper function is called it first gets the thread-local cache object for the function (creating if necessary). Then it calls that. You can do the same for fastcache as well - I'm just not sure how much that would slow things down.

**Comment by asmeurer** [comment]
I believe functools.lru_cache is thread-safe. If fastcache isn't, we should make fixes upstream to make it so. CC @pbrady. The compatibility version isn't needed any more as we are dropping Python 2 support.

**Comment by oscarbenjamin** [comment]
The `lru_cache` is thread-safe. The problem is that in the face of global flags like global evaluate cached sympy functions are not really pure functions any more. When changing the flag we have to clear the caches. That works fine in a single-threaded environment but when there are multiple threads accessing the same cache you can end up retrieving values stored in another thread that were computed using a different flag.

**Comment by pbrady** [comment]
The thread safety issues in fastcache were fixed in https://github.com/pbrady/fastcache/pull/48.  Some relevant discussion on fastcache performance here: https://github.com/pbrady/fastcache/issues/55.
Essentially, `functools.lru_cache` is an internal C module as of Python 3.5 and is actually faster than fastcache since it uses an internal dictionary api to avoid some of the hashing calls.  It may make sense to drop `fastcache` if you're targetting 3.5 and later.

**Comment by pbrady** [comment]
I opened an issue about the fastcache vs lru_cache performance a while ago #18070

**Comment by mcpl-sympy** [comment]
Thanks. Seems like I have had some misknowledges about this matter. Edited the comment.
